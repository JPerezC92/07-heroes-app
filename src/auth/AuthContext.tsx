import React, { createContext, useContext, useEffect, useReducer } from "react";
import { AuthAction, authReducer } from "./AuthReducer";

export interface IUser {
  logged: boolean;
  name: string | null;
}

interface IAuthContext extends IUser {
  dispatch: React.Dispatch<AuthAction>;
}

export const initialAuthContextState = {
  logged: false,
  name: null,
} as IAuthContext;

const AuthContext = createContext(initialAuthContextState);

const useAuthState = () => {
  const context = useContext(AuthContext);
  return context;
};

const init = (initialState: IUser) => {
  const userStorageString = localStorage.getItem("user");

  if (userStorageString) return JSON.parse(userStorageString) as IUser;

  return initialState;
};

const AuthContextProvider: React.FC = ({ children }) => {
  const [user, dispatch] = useReducer(
    authReducer,
    initialAuthContextState,
    init
  );

  const value: IAuthContext = { ...user, dispatch };

  useEffect(() => localStorage.setItem("user", JSON.stringify(user)), [user]);

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContextProvider, useAuthState };
