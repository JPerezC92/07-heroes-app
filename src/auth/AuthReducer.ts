import { IAuthContext } from "./AuthContext";

enum ActionType {
  login = "[auth] login",
  logout = "[auth] logout",
}

export const authReducer = (
  state: IAuthContext,
  action: { type: ActionType; payload: ObjectConstructor }
): IAuthContext => {
  switch (action.type) {
    case ActionType.login:
      return { ...action.payload, logged: true };

    case ActionType.logout:
      return { logged: false };

    default:
      return state;
  }
};
