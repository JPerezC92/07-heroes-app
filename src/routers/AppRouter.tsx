import { BrowserRouter as Router, Switch } from "react-router-dom";
import { useAuthState } from "../auth/AuthContext";
import LoginScreen from "../components/login/LoginScreen";
import DashboardRoutes from "./DashboardRoutes";
import PrivateRoute from "./PrivateRoute";
import PublicRoute from "./PublicRoute";

const AppRouter = () => {
  const { user } = useAuthState();

  return (
    <Router>
      <div>
        <Switch>
          <PublicRoute
            exact
            isAuthenticated={user.logged}
            path="/login"
            Component={LoginScreen}
          />
          <PrivateRoute
            isAuthenticated={user.logged}
            path="/"
            Component={DashboardRoutes}
          />
        </Switch>
      </div>
    </Router>
  );
};

export default AppRouter;
