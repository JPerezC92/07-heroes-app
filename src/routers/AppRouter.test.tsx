import { mount } from "enzyme";
import { useAuthState } from "../auth/AuthContext";
import AppRouter from "./AppRouter";

jest.mock("../auth/AuthContext", () => {
  return {
    useAuthState: jest.fn(),
  };
});

describe("Test on <AppRouter />", () => {
  test("should render login if user isn't authenticated", () => {
    (useAuthState as jest.MockedFunction<any>).mockReturnValue({
      user: { logged: false, name: "null" },
    });
    const wrapper = mount(<AppRouter />);

    expect(wrapper).toMatchSnapshot();
  });

  test("should render marvel  if user is authenticated", () => {
    const name = "Philip";

    (useAuthState as jest.MockedFunction<any>).mockReturnValue({
      user: { logged: true, name },
    });
    const wrapper = mount(<AppRouter />);

    expect(wrapper.contains(name)).toBeTruthy();
    expect(wrapper.find(".navbar").exists()).toBeTruthy();
  });
});
