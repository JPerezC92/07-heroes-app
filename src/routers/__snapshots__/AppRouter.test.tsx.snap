// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <AppRouter /> should render login if user isn't authenticated 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      Login Screen
    </h1>
    <hr />
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Login
    </button>
  </div>
</div>
`;
